hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/RSRpcServices.java

```diff
@@ -3616,8 +3620,9 @@ public ScanResponse scan(final RpcController controller, final ScanRequest reque
         if (region.getCoprocessorHost() != null) {
           Boolean bypass = region.getCoprocessorHost().preScannerNext(scanner, results, rows);
           if (!results.isEmpty()) {
+            Object lastBlock = null;
             for (Result r : results) {
-              lastBlock.setValue(addSize(rpcCall, r, lastBlock.getValue()));
+              lastBlock = addSize(rpcCall, r, lastBlock);
```

Comment:
Right, so these `results` here come from a coprocessor. We still want to count the size of those, and the coprocessor does not have access to the scanner context.

The way addSize works is it keeps track of the last block and only increments if `block != lastBlock`. This approach is ok for the coprocessor results.

Previously the `lastBlock` value was kept and then passed into the downstream `scan(..)` method in the MutableObject.  Then where you commented before (line 3401) we would call addSize again, and the passed in `lastBlock` value would be used for that equality check.

Now that we use ScannerContext for the accounting of the actual scans, we no longer need the addSize call on line 3401. So I removed that and the MutableObject.

But we still need lastBlock for checking the coprocessor results, since the way addSize works is to keep track of a lastBlock across invocations. `addSize` is also still used by Get/Multiget (though I hope to unify those in a later issue).

Reply:
I'm still a bit confused, if we do not pass lastBlock done, how could we count the size of these blocks and limit the total size?

Reply:
Ok I see what you're saying. Yes these blocks are not counted towards the max size. Just counted for reporting metrics and slow log threshold. 

These are two separate data sets, but I can link them. When creating the maxQuotaResultSize, I can subtract rpcCall.getResponseBlockSize(). That will account for the blocks accumulated in the coprocessor. 

I won't need lastBlock. We can't really use that below now that we use scanner context which is far more accurate than addSize estimation. We could theoretically try piping it through all the way into HFileReaderImpl where we record block sizes. But I don't think it's worth it for one block overlap. 

I will make the change to subtract counter block size when setting limits on ScannerContext.

Reply:
I just pushed fixes, and added comments about what's happening with addSize and ScannerContext